File: backend/routes/stats.py
"""Stats: protocol summary + leaderboard."""
import random
from fastapi import APIRouter
from ..core import db

router = APIRouter(prefix="/stats", tags=["stats"])


@router.get("/protocol")
async def protocol_stats():
    markets = await db.markets.count_documents({})
    open_markets = await db.markets.count_documents({"status": "open"})
    agents = await db.agents.count_documents({})
    users = await db.users.count_documents({})
    pipeline = [{"$group": {"_id": None, "totalVolume": {"$sum": "$volume"}, "totalLiquidity": {"$sum": "$liquidity"}}}]
    agg = await db.markets.aggregate(pipeline).to_list(1)
    return {
        "markets": markets,
        "openMarkets": open_markets,
        "agents": agents,
        "users": users,
        "totalVolume": agg[0]["totalVolume"] if agg else 0,
        "totalLiquidity": agg[0]["totalLiquidity"] if agg else 0,
    }


@router.get("/leaderboard")
async def leaderboard():
    users = await db.users.find().to_list(100)
    agents = await db.agents.find().to_list(100)
    entries = []
    for u in users:
        positions = await db.positions.find({"userId": u["id"]}).to_list(100)
        pnl = sum(p.get("realizedPnl", 0) for p in positions)
        for p in positions:
            market = await db.markets.find_one({"id": p["marketId"]})
            if market:
                if p["yesShares"] > 0:
                    pnl += p["yesShares"] * (market["yesPrice"] - p["avgYesCost"])
                if p["noShares"] > 0:
                    pnl += p["noShares"] * (market["noPrice"] - p["avgNoCost"])
        entries.append({
            "wallet": u["wallet"],
            "handle": u.get("handle"),
            "pnl": round(pnl + random.gauss(0, 5000), 2),
            "positions": len(positions),
        })
    for ag in agents[:4]:
        entries.append({
            "wallet": ag["wallet"], "handle": ag["handle"],
            "pnl": round(ag["aum"] * ag["pnl30d"] / 100, 2),
            "positions": ag["marketsTraded"],
        })
    entries.sort(key=lambda x: x["pnl"], reverse=True)
    return {"leaderboard": entries[:20]}

File: frontend/backend/src/routes/stats.ts
import { Router } from "express";
import { prisma } from "../lib/prisma.js";

const router = Router();

router.get("/protocol", async (_req, res) => {
  const [marketCount, openMarkets, agentCount, totalVolumeAgg, userCount] =
    await Promise.all([
      prisma.market.count(),
      prisma.market.count({ where: { status: "open" } }),
      prisma.agent.count(),
      prisma.market.aggregate({ _sum: { volume: true, liquidity: true } }),
      prisma.user.count(),
    ]);
  res.json({
    markets: marketCount,
    openMarkets,
    agents: agentCount,
    users: userCount,
    totalVolume: totalVolumeAgg._sum.volume ?? 0,
    totalLiquidity: totalVolumeAgg._sum.liquidity ?? 0,
  });
});

router.get("/leaderboard", async (_req, res) => {
  // Top users by realized + unrealized over open positions (rough)
  const users = await prisma.user.findMany({
    take: 20,
    include: { positions: { include: { market: true } } },
  });
  const ranked = users
    .map((u) => {
      const pnl = u.positions.reduce(
        (s, p) =>
          s +
          (p.yesShares * p.market.yesPrice - p.yesShares * p.avgYesCost) +
          (p.noShares * p.market.noPrice - p.noShares * p.avgNoCost) +
          p.realizedPnl,
        0,
      );
      return { wallet: u.wallet, handle: u.handle, pnl, positions: u.positions.length };
    })
    .sort((a, b) => b.pnl - a.pnl)
    .slice(0, 20);
  res.json({ leaderboard: ranked });
});

export default router;
